src/v/kafka/server/requests.cc

```diff
@@ -179,22 +182,23 @@ handle_auth(request_context&& ctx, ss::smp_service_group g) {
     }
 }
 
-ss::future<response_ptr>
+process_result_stages
 process_request(request_context&& ctx, ss::smp_service_group g) {
```

Comment:
another simplification might be to redefine request processing as having a foreground and background phase of execution. the same request context would be passed to both phases. this would probably simplify the code that arranges for both futures to be available as a single return value from the produce handler. it's just a thought, though. it might make sense to do this if we end up wanting a staged execution for a couple other request types.

Reply:
I think I do not fully understand the first comment. If we would still use `process_request` in `connection_context` it would not allow us to leverage the two stage processing.

Reply:
the second comment suggests a more a more fundamental change. like, imagine if `producer_handler::handle` were split into `produce_handler::prepare(ctx)` and `produce_handler::finish(ctx)`. I'm claiming that would simplify a lot, but would be a big change. we don't need to tackle it right now.

Reply:
This is great idea, I really like it, maybe we can refactor this in next step.

Reply:
I was thinking on how to deal with this in a neat way. Approach with two methods is great although we would require to pass some state across them in `request_context`